connaisseur/sigstore_validator.py

```diff
@@ -0,0 +1,93 @@
+import json
+import logging
+import re
+import subprocess  # nosec
+
+from connaisseur.crypto import decode_and_verify_ecdsa_key
+from connaisseur.exceptions import (
+    CosignError,
+    CosignTimeout,
+    NotFoundException,
+    ValidationError,
+    UnexpectedCosignData,
+)
+
+
+def get_cosign_validated_digests(image: str, pubkey: str):
+    """
+    Gets and processes cosign validation output for a given `image` and `pubkey`
+    and either returns a list of valid digests or raises a suitable exception
+    in case no valid signature is found or cosign fails.
+    """
+    returncode, stdout, stderr = invoke_cosign(image, pubkey)
+    logging.info(
+        "COSIGN output for image: %s; RETURNCODE: %s; STDOUT: %s; STDERR: %s",
+        image,
+        returncode,
+        stdout,
+        stderr,
+    )
+    digests = []
+    if returncode == 0:
+        for sig in stdout.splitlines():
+            try:
+                sig_data = json.loads(sig)
+                try:
+                    digest = sig_data["Critical"]["Image"].get(
+                        "Docker-manifest-digest", ""
+                    )
+                    if re.match(r"(sha256:)?[0-9A-Fa-f]{64}", digest) is None:
+                        raise Exception(
+                            "digest does not match expected digest pattern."
+                        )
+                except Exception as err:
+                    raise UnexpectedCosignData(
+                        f"could not retrieve valid digest from data received by cosign: {err}"
+                    ) from err
+
+                # remove prefix 'sha256' in case it exists
+                digests.append(digest[7:] if digest[:7] == "sha256:" else digest)
+            except json.JSONDecodeError:
+                logging.info("Non-json signature data from Cosign: %s", sig)
+                pass
```

Comment:
This `except` clause should be right after `json.loads` (line 34), right? The (currently) inner `try`-`except` block could then go up one level, streamlining the whole thing.

Reply:
I just tried but noticed that we need `sig_data` in the inner block which is not defined if a `JSONDecodeError` is thrown (and caught by us).